data_structures/skip_list.cpp

```diff
@@ -0,0 +1,212 @@
+/**
+ * A skip list is a data structure that is used for storing a sorted list of items with a
+ *  help of hierarchy of linked lists that connect increasingly sparse subsequences of the items
+ * 
+ *  References used: GeeksForGeeks skip list code, https://iq.opengenus.org/skip-list/ Pseudo Code.
+*/
+
+#include <iostream>
+#include <vector>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <time.h>   
+#include <cstdlib>
+
+// using namespace std; 
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2
+#define P 0.5
+  
+
+/**
+ *  Node structure [Key][Node*, Node*...]
+*/
+class Node { 
+public: 
+    int key; 
+    void* value;
+    /*Forward Array*/
+    vector<Node*> forward;
+    Node(int key, int level, void* value); 
+};
+
+  
+Node::Node(int key, int level, void* value) { 
+    this->key = key; 
+  
+    /*Initialization of forward vector*/
+    for (int i = 0; i < sizeof(Node*)*(level+1); i++){
+        forward.push_back(NULL);
+    }
+} 
+  
+// Class for Skip list 
+class SkipList { 
+    int level; 
+    Node *header; 
+public: 
+    SkipList(); 
+    int randomLevel(); 
+    void insertElement(int, void*); 
+    void deleteElement(int); 
+    void* searchElement(int); 
+    void displayList(); 
+    ~SkipList();
+};
+
+
+  
+SkipList::SkipList() { 
+    level = 0; 
+    /* Header initialization*/
+    header = new Node(-1, MAXLVL, NULL); 
+} 
+
+
+SkipList::~SkipList(){
+    delete header;
+    for(int i=0; i <= level; i++) { 
+        Node *node = header->forward[i]; 
```

Comment:
by here, `header` variable is already deleted and invalid. the `header` variable should be deleted at the end, after the loop.

Reply:
yeah correcto, deleting header should happen afterwards